.github/workflows/reusable_release_automation.yaml

```
@@ -0,0 +1,104 @@
+
+on:
+  workflow_call:
+    inputs:
+      block_endpoint:
+        description: This endpoint should respond status 200 for the promotion to continue.
+        required: false
+        type: string
+        default: https://newrelic.github.io/coreint-automation/automatic_release_enable
+    secrets:
+      COREINT_BOT_TOKEN:
+        required: true
+      COREINT_SLACK_CHANNEL:
+        required: true
+      COREINT_SLACK_TOKEN:
+        required: true
+
+jobs:
+  check_block_endpoint:
+    name: Check block endpoint
+    runs-on: ubuntu-latest
+    steps:
+      - name: Check block endpoint
+        run: |
+          http_status=$(curl -s -o /dev/null -w "%{http_code}" "${{ inputs.block_endpoint }}")
+
+          if [ "$http_status" != 200 ]; then
+            echo "Fail checking release block endpoint: ${{ inputs.block_endpoint }}"
+            exit 1
+          fi
+
+  promote_prerelease:
+    needs: check_block_endpoint
+    runs-on: ubuntu-latest
+    steps:
+      - name: Promote previous prerelease to release
+        env:
+          GH_TOKEN: "${{ secrets.COREINT_BOT_TOKEN }}"
+        run: |
+          # Get the date of the latest release.
+          LATEST_RELEASE_DATE=$(
+            gh release list --json publishedAt,isLatest -R "${{ github.repository }}" --jq '
+              .[] | select(.isLatest) | .publishedAt | fromdateiso8601
+            '
+          )
+
+          # List all releases, filter by prereleases that were made later than the latest release, sort by date,
+          # reverse because the last one is the most recent and only grab the last prerelease
+          NON_PUBLISHED_PRERELEASE=$(
+            gh release list --json name,publishedAt,isPrerelease,tagName -R "${{ github.repository }}" --order desc --jq '
+              [
+                  .[] | select( .isPrerelease and ((.publishedAt | fromdateiso8601) > '${LATEST_RELEASE_DATE}'))
+              ]
+              | sort_by(.publishedAt |= fromdateiso8601) | first
+            '
+          )
+
+          # Extract the name and tag from the latest prerelease
+          NON_PUBLISHED_PRERELEASE_NAME=$(jq -r '.name' <<<"$NON_PUBLISHED_PRERELEASE")
+          NON_PUBLISHED_PRERELEASE_TAG=$(jq -r '.tagName' <<<"$NON_PUBLISHED_PRERELEASE")
+
+          # Check if the prerelease name matches its tag and promote if they are the same. In case the prerelase is labeled
+          # with "artifacts-pending" or "pre-release-failure" suffixes, it will fail.
+          if [[ "$NON_PUBLISHED_PRERELEASE_NAME" == "$NON_PUBLISHED_PRERELEASE_TAG" ]]; then
+            # Promote the prerelease to a full release
+            gh release edit "$NON_PUBLISHED_PRERELEASE_TAG" --repo "${{ github.repository }}" --draft=false --prerelease=false
+          else
+            # Do not promote and set to notify
+            echo "Prerelease $NON_PUBLISHED_PRERELEASE_NAME is blocked since it does not match with $NON_PUBLISHED_PRERELEASE_TAG"
+            exit 1
+          fi
```

```suggestion
          if [[ "$NON_PUBLISHED_PRERELEASE_NAME" != "$NON_PUBLISHED_PRERELEASE_TAG" ]]; then
            echo "Prerelease $NON_PUBLISHED_PRERELEASE_NAME is blocked since it does not match with $NON_PUBLISHED_PRERELEASE_TAG"
            exit 1
          fi

          gh release edit "$NON_PUBLISHED_PRERELEASE_TAG" --repo "${{ github.repository }}" --draft=false --prerelease=false
```

Now I am regretting the change I suggested...

I am wondering if we really want to abort the new prerelease if we are not able to promote the latest prerelease to release... The philosophy we have right now is just to keep the ball rolling. Instead of aborting everything, just let the new prerelease to pop and next wee the next prerelease will be promoted.

With the approach we have now, if a prerelease failed, it will not promote (that is expected and OK) but will stop a new prerelease. Is it really what we want?

---

Even if you revert, the prerelease is not going to trigger because of this:
https://github.com/newrelic/coreint-automation/blob/5b1e195d56fd635470f8c62872dc27237d5531b6/.github/workflows/reusable_release_automation.yaml#L96-L99

The `needs` field need something different, something like:
```yaml
  trigger_prerelease:
    needs: promote_prerelease
    if: ${{ always() }}
    uses: ./.github/workflows/reusable_trigger_prerelease.yaml
    secrets:
```

So the job that trigger a prerelease is executed regardless if the needed job failed or not.

Do we also want to move the slack notification out of the release promotion job?

---

> Do we also want to move the slack notification out of the release promotion job?

There's another notify failure slack step inside the `reusable_trigger_prerelease`. That's why I kept it here and I didn't refactor that reusable since for some integrations we might need to still call it directly.